pinery-lims/src/main/java/ca/on/oicr/pinery/lims/DefaultSampleProvenance.java

```diff
@@ -466,4 +465,30 @@ private HashMultimap<String, String> processSampleAttributes(Sample sample) {
 
     return attrs;
   }
+
+  @Override
+  public List<String> getBatchIds() {
+    // Should return an empty list if there are none. Null is used to omit the attribue from old
+    // provenance versons (see SampleProvenanceDto and SimpleSampleProvenance)
+    List<String> batchIds = new ArrayList<>();
+    addBatchId(sample, batchIds);
+    if (parentSamples != null) {
+      for (Sample parent : parentSamples) {
+        addBatchId(parent, batchIds);
+      }
+    }
+    return batchIds;
+  }
+
+  private static void addBatchId(Sample sample, List<String> list) {
+    if (sample.getAttributes() == null) {
+      return;
+    }
+    Attribute attribute = sample.getAttributes().stream()
```

Comment:
It may be good to sort the ids to ensure object hashes don't change between invocations.

Reply:
Good point. There should only be one batch ID per sample, but the Pinery API doesn't guarantee that. Sorted now